### ISSUE TITLE
When using cgroups, system memory can be negative

### ISSUE THREAD

**Issue by jmcphers** [rstudio/rstudio#9353]
### System details

    RStudio Edition : Server
    RStudio Version : 1.4.1707

### Steps to reproduce the problem

Unknown, but currently only see in Kubernetes with `cgroups`. Open Tools -> Memory Usage and observe the following report claiming that system memory used is negative:

![image](https://user-images.githubusercontent.com/470418/118298318-2d1d8180-b494-11eb-8a2a-4121d13c3fce.png)

### Describe the problem in detail

The problem is that we are reading the `/proc` filesystem to determine the process's memory usage..

https://github.com/rstudio/rstudio/blob/b243985d31bb50ab59595de1e6babf1f6691406f/src/cpp/core/system/LinuxResources.cpp#L60-L61

In R, you can see that this tells us that the process's resident size is 169,484 KiB, based on a page size of 4 KiB.

```
> readLines("/proc/self/statm")
[1] "328572 42371 15009 5060 0 94377 0"
> (42371 * 4096) / 1024
[1] 169484
```

We get our system memory usage in bytes from cgroups:

https://github.com/rstudio/rstudio/blob/b243985d31bb50ab59595de1e6babf1f6691406f/src/cpp/core/system/LinuxResources.cpp#L214-L216

Again, simulating this in R:

```
> readLines("/sys/fs/cgroup/memory/memory.usage_in_bytes")
[1] "138833920"
> 138833920 / 1024
[1] 135580
```

So ... even though there is a process with a resident size of 169,484 KiB, the total system memory usage is only 135,580 KiB. We must not be comparing apples to apples here.

### Describe the behavior you expected

System memory should never be negative.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [x] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by jmcphers** [comment]
https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt 

> For efficiency, as other kernel components, memory cgroup uses some optimization to avoid unnecessary cacheline false sharing. usage_in_bytes is affected by the method and doesn't show 'exact' value of memory (and swap) usage, it's a fuzz value for efficient access. (Of course, when necessary, it's synchronized.) If you want to know more exact memory usage, you should use RSS+CACHE(+SWAP) value in memory.stat(see 5.2).

So maybe we need to read more fine-grained values from `memory.stat` since `usage_in_bytes` is fuzzy.

**Comment by atheriel** [comment]
The root cause of this is that it is not generally possible to answer a question like "how much memory is a process using?" in a straightforward way on Linux, because Linux caches the hell out of anything processes might be sharing to improve performance. (See Viacheslav Biriukov's [excellent series on the Linux Page Cache](https://biriukov.dev/docs/page-cache/7-how-much-memory-my-program-uses-or-the-tale-of-working-set-size/#how-much-memory-my-program-uses-or-the-tale-of-working-set-size) for more.)

Memory `Used by session` is currently measured by reading a value -- specifically `VmRSS` -- from `/proc/self/statm`. From the `proc` man page:

> VmRSS: Resident set size. Note  that the value here is the sum of RssAnon, RssFile, and RssShmem...
>
> RssAnon: Size of resident anonymous memory...
>
> RssFile: Size of resident file mappings...
>
> RssShmem: Size of resident shared memory (includes  System  V  shared  memory,  mappings  from tmpfs(5), and shared anonymous mappings).

Unfortunately, `RssShmem` is by definition shared across multiple processes, as is anything in `RssFile` that ends up in the page cache. So if you add up the `VmRSS` value across all processes, you will considerably exceed actual system memory. Only `RssAnon` is ([mostly](https://stackoverflow.com/a/66593890)) scoped to an individual process.

Memory `Used by system` is determined differently depending on whether we detect that we're inside a memory-restrictive cgroup or not (leaving aside that on any systemd-based distribution we are *always* inside a cgroup).

When we're _not_ in a restrictive cgroup, we check `/proc/meminfo` for total and available system memory and compute system usage as the difference, minus the `Used by session` value above. This is problematic because `RssShmem` and probably most of `RssFile` are really shared across processes and won't go away when your session does, so we're really overestimating the memory usage of a session here.

When we _are_ in a restrictive cgroup, we use `memory.usage_in_bytes`. What does that actually contain? From the kernel's docs on [cgroup Memory Ownership](https://docs.kernel.org/admin-guide/cgroup-v2.html#memory-ownership):

> A memory area is charged to the cgroup which instantiated it [..] A memory area may be used by processes belonging to different cgroups. To which cgroup the area will be charged is [sic] in-deterministic

In other words, the overestimation due to including `RssShmem` and `RssFile` is now much, much worse, because we are counting memory shared among processes in other cgroups and attributing it to our session, while at the same time scoping the "system" to a single cgroup. In this light, it's not surprising to see a negative `Used by system` value.

Finally, with the current implementation, the meaning of "system" and "available" memory is **completely different** inside and outside a cgroup:

* Outside a cgroup we show `MemAvailable` from `/proc/meminfo` -- which is a loose estimate of actual memory available on that machine for use by a program like the session -- and attribute the remainder to the "system".

* Inside a cgroup we do the opposite: we estimate "system" memory from current in-cgroup usage and imply to users that the remainder (up to the cgroup memory limit) *is available for use by the session*. But there's no reason to believe that this is generally true: only in the narrow case of "[on Kubernetes with a memory request equal to a memory limit](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed)" will that amount of memory be truly available for a session. On a more traditional cgroup setup -- say, when running RStudio or Workbench as a unit under systemd, which uses cgroups -- there may be a high or unrestricted memory limit, but that doesn't mean an unlimited amount of memory is available.

In sum: I don't really think we can fix this by making `Used by session` a more apples-to-apples comparison inside a cgroup to `VmRSS` from `/proc/self/statm`.

Instead, I think we would need a much more careful estimate of memory used and available inside a container, perhaps with a different UI.

**Comment by cgates** [comment]
To validate comment above, I just updated 6/6/2025 and am seeing negative memory for the first time. 

```
RStudio 2025.05.1+513 "Mariposa Orchid" Release (ab7c1bc795c7dcff8f26215b832a3649a19fc16c, 2025-06-01) for Ubuntu Jammy
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.5 Safari/605.1.15, Quarto 1.6.42
```

After some memory intensive ops, I see this:
<img width="612" alt="Image" src="https://github.com/user-attachments/assets/1fa4d1f6-a1c2-4060-b78e-39785e7f14e2" />

These memory stats persist will persist until the rstudio-server is restarted, and will show for any subsequent user session. Below is a different user with a completely fresh session starting out using almost 9Gb. But not really - because note the discrepancy of the reported memory used by the session between RStudio and R below.

<img width="1226" alt="Image" src="https://github.com/user-attachments/assets/5db5d66e-7181-48c6-9293-03f170ba40f0" />

```
> pryr::mem_used()
57.2 MB
```

Any ideas?

**Comment by jeffvroom** [comment]
Yes, this is a report of an old bug that is now a similar new bug introduced in 2025.05 that we should fix. The code used to use just the session process RSS size, but is now using the cgroup's "usage_in_bytes" to represent the session's memory. Since usage_in_bytes includes non-resident memory, but total and free memory do not, the equation: Used-by-system = (total - free) - session memory isn't accurate. We should probably have the system collect both the cgroup memory, to use in memory limit checks, and the RSS size, to be used to see how much resident memory it's using.

**Comment by cgates** [comment]
FWIW, this behavior also shows up in the latest daily:

```
RStudio 2025.08.0-daily+195 "Cucumberleaf Sunflower" Daily (e6c430abfed8c93d5dd9f5ca55849cd5d72fc37e, 2025-06-10) for Ubuntu Jammy
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.5 Safari/605.1.15, Quarto 1.7.31 (/usr/lib/rstudio-server/bin/quarto-1.6.39/bin/quarto)
```

<img width="547" alt="Image" src="https://github.com/user-attachments/assets/a2bedb36-380c-4c72-8bc0-4929f2c11a09" />
